### PR TITLE
WIP: typescript love on the tests

### PR DESCRIPTION
**Pull request by dwelch2344**

### CODE REVIEW

File: .flowconfig [hunk 1]
```diff
@@ -0,0 +1,23 @@
+[ignore]
```
Comment (pikax) [inline .flowconfig:1]: This file should have been removed on my branch 🤔

File: test/helpers/to-have-been-warned.ts [hunk 2]
```diff
@@ -50,8 +51,12 @@ beforeEach(() => {
   spyOn(console, 'warn')
   spyOn(console, 'error')
   jasmine.addMatchers({
-    toHaveBeenWarned: () => createCompareFn(console.error),
-    toHaveBeenTipped: () => createCompareFn(console.warn)
+    toHaveBeenWarned: () => 
+      // @ts-ignore
```
Comment (pikax) [inline test/helpers/to-have-been-warned.ts]: can you use `//@ts-expect-error` instead of `@ts-ignore` pls?

File: test/helpers/trigger-event.ts [hunk 3]
```diff
@@ -1,8 +1,9 @@
-window.triggerEvent = function triggerEvent (target, event, process) {
+global.triggerEvent = function triggerEvent (target, event, process) {
```
Comment (pikax) [inline test/helpers/trigger-event.ts]: let's keep `window`, we should avoid changing javascript code, "improvements" should be handled at an advanced phase.

File: test/helpers/wait-for-update.ts [hunk 4]
```diff
@@ -21,15 +21,15 @@ window.waitForUpdate = initialCb => {
     if (queue.length) {
       let hasError = false
       try {
-        job.wait ? job(shift) : job()
+        job.wait ? job(shift) : (job ? job() : null)
```
Comment (pikax) [inline test/helpers/wait-for-update.ts]: let's not make any javascript changes for now, if the object can be `undefined` let's add an `@ts-expect-error` and we can do a mass processing later on

**Review by pikax — COMMENTED** [review]
Just an initial review, main point is to not change javascript code, please reach out to me if you have questions on a specific code

File: packages/vue-server-renderer/basic.js [hunk 5]
```diff
@@ -2953,4677 +2952,4677 @@
...
+              const tips = [];
+              let warn = (msg, range, tip) => {
+                  (tip ? tips : errors).push(msg);
+              };
+              if (options) {
+                  if (options.outputSourceRange) {
+                      // $flow-disable-line
+                      const leadingSpaceLength = template.match(/^\s*/)[0].length;
+                      warn = (msg, range, tip) => {
+                          const data = typeof (msg) === 'string' ? { msg } : msg;
```
Comment (pikax) [inline packages/vue-server-renderer/basic.js]: rollback this change and similar pls 🙏

File: packages/vue-template-compiler/browser.js [hunk 6]
```diff
@@ -2240,2695 +2240,2695 @@
...
+	        function compile(template, options) {
+	            const finalOptions = Object.create(baseOptions);
+	            const errors = [];
+	            const tips = [];
+	            let warn = (msg, range, tip) => {
+	                (tip ? tips : errors).push(msg);
+	            };
+	            if (options) {
+	                if (options.outputSourceRange) {
+	                    // $flow-disable-line
```
Comment (pikax) [inline packages/vue-template-compiler/browser.js]: maybe we can find and replace `// $flow-disable-line`  from all the code base, it's not needed

File: test/e2e/specs/svg.ts [hunk 7]
```diff
@@ -1,4 +1,8 @@
-/* globals stats, valueToPoint */
+/* globals */
```
Comment (pikax) [inline test/e2e/specs/svg.ts]: we don't need this comment anymore :)

File: test/helpers/wait-for-update.ts [hunk 8]
```diff
@@ -12,6 +12,7 @@ import Vue from 'vue'
 //   // more assertions...
 // })
 // .then(done)
+// @ts-expect-error
 window.waitForUpdate = initialCb => {
```
Comment (pikax) [inline test/helpers/wait-for-update.ts]: Here we should be able to add `waitForUpdate` on the window object, removing the need to `//@ts-expect-error`

File: test/helpers/wait-for-update.ts [hunk 9]
```diff
@@ -29,7 +30,7 @@ window.waitForUpdate = initialCb => {
           done.fail(e)
         }
       }
-      if (!hasError && !job.wait) {
+      if (!hasError && !job?.wait) {
```
Comment (pikax) [inline test/helpers/wait-for-update.ts]:
```suggestion
      if (!hasError && !job!.wait) {
```
Just to keep javascript code unchanged (also the other similar changes) pls

File: test/ssr/compile-with-webpack.ts [hunk 10]
```diff
@@ -3,6 +3,9 @@ import webpack from 'webpack'
 import MemoryFS from 'memory-fs'
 
 export function compileWithWebpack (file, extraConfig, cb) {
+  if( process.env.FOOBAR !== 'blah') {
+    throw new Error('die')
+  }
```
Comment (pikax) [inline test/ssr/compile-with-webpack.ts]: remove this :D

File: test/unit/features/component/component-scoped-slot.spec.ts [hunk 11]
```diff
@@ -508,28 +508,28 @@ describe('Component scoped slot', () => {
...
-            return <div>
-              {this.$scopedSlots.default({ msg: this.msg })}
-            </div>
-          }
-        }
-      }
-    }).$mount()
-    expect(vm.$el.innerHTML).toBe('<span>hello</span>')
-  })
+  // it('render function usage (JSX)', () => {
```
Comment (pikax) [inline test/unit/features/component/component-scoped-slot.spec.ts:511]: Why this is commented?

File: test/unit/features/debug.spec.ts [hunk 12]
```diff
@@ -3,7 +3,7 @@ import { formatComponentName, warn } from 'core/util/debug'
 
 describe('Debug utilities', () => {
   it('properly format component names', () => {
-    const vm = new Vue()
+    const vm = new Vue('div')
```
Comment (pikax) [inline test/unit/features/debug.spec.ts]: why passing `'div'`? it should work without it (javascript wise), if this is typescript error, please add a `@ts-expect-error` with a comment pls

File: test/unit/features/error-handling.spec.ts [hunk 13]
```diff
@@ -170,12 +170,12 @@ describe('Error handling', () => {
     Vue.nextTick(() => {
       expect(spy).toHaveBeenCalledWith(err1, undefined, 'nextTick')
 
-      const vm = new Vue()
+      const vm = new Vue('div')
```
Comment (pikax) [inline test/unit/features/error-handling.spec.ts]: here too pls

File: test/unit/features/global-api/mixin.spec.ts [hunk 14]
```diff
@@ -19,8 +19,8 @@ describe('Global API: mixin', () => {
   })
 
   it('should work for constructors created before mixin is applied', () => {
-    const calls = []
-    const Test = Vue.extend({
+    const calls: any[] = []
+    const Test: Vue = Vue.extend({
```
Comment (pikax) [inline test/unit/features/global-api/mixin.spec.ts:23]: do you need `: Vue` I would expect `Vue.extend ` to return `Vue`.

If this is necessary, don't need to change

File: test/unit/features/instance/init.spec.ts [hunk 15]
```diff
@@ -2,11 +2,11 @@ import Vue from 'vue'
 
 describe('Initialization', () => {
   it('without new', () => {
-    try { Vue() } catch (e) {}
+    try { Vue('div') } catch (e) {}
```
Comment (pikax) [inline test/unit/features/instance/init.spec.ts]: also this

File: test/unit/features/instance/methods-events.spec.ts [hunk 16]
```diff
@@ -3,7 +3,7 @@ import Vue from 'vue'
 describe('Instance methods events', () => {
   let vm, spy
   beforeEach(() => {
-    vm = new Vue()
+    vm = new Vue('div')
```
Comment (pikax) [inline test/unit/features/instance/methods-events.spec.ts]: also here

File: test/unit/features/instance/properties.spec.ts [hunk 17]
```diff
@@ -161,7 +161,7 @@ describe('Instance properties', () => {
   })
 
   it('warn mutating $attrs', () => {
-    const vm = new Vue()
+    const vm = new Vue('div')
```
Comment (pikax) [inline test/unit/features/instance/properties.spec.ts]: also here

File: test/unit/features/options/lifecycle.spec.ts [hunk 18]
```diff
@@ -78,7 +78,7 @@ describe('Options lifecycle hooks', () => {
 
     // #3898
     it('should call for manually mounted instance with parent', () => {
-      const parent = new Vue()
+      const parent = new Vue('div')
```
Comment (pikax) [inline test/unit/features/options/lifecycle.spec.ts]: also here

File: test/unit/features/options/render.spec.ts [hunk 19]
```diff
@@ -33,7 +33,7 @@ describe('Options render', () => {
   })
 
   it('should warn non `render` option and non `template` option', () => {
-    new Vue().$mount()
+    new Vue('div').$mount()
```
Comment (pikax) [inline test/unit/features/options/render.spec.ts]: also here

File: test/unit/features/transition/transition.spec.ts [hunk 20]
```diff
@@ -23,7 +23,7 @@ if (!isIE9) {
       // should not apply transition on initial render by default
       expect(vm.$el.innerHTML).toBe('<div class="test">foo</div>')
       vm.ok = false
-      waitForUpdate(() => {
+      global.waitForUpdate(() => {
```
Comment (pikax) [inline test/unit/features/transition/transition.spec.ts]: global should be needed here

File: test/unit/modules/observer/observer.spec.ts [hunk 21]
```diff
@@ -14,7 +14,7 @@ describe('Observer', () => {
     const ob1 = observe(1)
     expect(ob1).toBeUndefined()
     // avoid vue instance
-    const ob2 = observe(new Vue())
+    const ob2 = observe(new Vue('div'))
```
Comment (pikax) [inline test/unit/modules/observer/observer.spec.ts]: div here

**Review by pikax — CHANGES_REQUESTED** [review]
Just a few changes, apologies for so many comments, but was to be able to know what files would need to be updated, thank you very much :)

File: test/unit/modules/observer/observer.spec.ts (hunk 21 above)
Comment (dwelch2344) [reply test/unit/modules/observer/observer.spec.ts]: Do we want to roll this back? Contract is unhappy with no-arg constructor, but maybe that's part of the inconsistency you found in the API. 

Either way, LMK and I'll get it done across the board :)

Comment (pikax) [reply test/unit/modules/observer/observer.spec.ts]: I think we should rollback, it was working before 🤔 
We can make args to be optional

File: test/ssr/compile-with-webpack.ts (hunk 10 above)
Comment (dwelch2344) [reply test/ssr/compile-with-webpack.ts]: Hahah oh yeah, forgot I got desperate while picking apart the builds ;)

File: test/unit/features/global-api/mixin.spec.ts (hunk 14 above)
Comment (dwelch2344) [reply test/unit/features/global-api/mixin.spec.ts:23]: Yeah, typescript wasn't happy for every instance of this, but I didn't want to dig into the typing files just yet. I'll leave it for now, but putting a mental TODO

File: test/unit/modules/observer/observer.spec.ts (hunk 21 above)
Comment (dwelch2344) [reply test/unit/modules/observer/observer.spec.ts]: Nevermind, saw the other comments. Cleaned!

**Review by pikax — APPROVED** [review]

**Pull request merged**